WcaOnRails/app/webpacker/components/RolesTab/index.jsx

```diff
@@ -6,77 +6,133 @@ import {
   List,
 } from 'semantic-ui-react';
 import useLoadedData from '../../lib/hooks/useLoadedData';
+import useSaveAction from '../../lib/hooks/useSaveAction';
 import {
   roleListUrl,
+  currentUserUrl,
+  teamUrl,
 } from '../../lib/requests/routes.js.erb';
 import Errored from '../Requests/Errored';
 import Loading from '../Requests/Loading';
 import RoleForm from './RoleForm';
+import I18n from '../../lib/i18n';
 
-function RoleFormModal({
-  trigger, userId, isActiveRole, sync,
-}) {
-  const [open, setOpen] = React.useState(false);
-
-  return (
-    <Modal
-      size="fullscreen"
-      onClose={() => {
-        setOpen(false);
-        sync();
-      }}
-      onOpen={() => setOpen(true)}
-      open={open}
-      trigger={trigger}
-    >
-      <Modal.Content>
-        <RoleForm
-          userId={userId}
-          isActiveRole={isActiveRole}
-        />
-      </Modal.Content>
-    </Modal>
-  );
-}
+// let i18n-tasks know the key is used
+// i18n-tasks-use t('enums.user.role_status.trainee_delegate')
+// i18n-tasks-use t('enums.user.role_status.candidate_delegate')
+// i18n-tasks-use t('enums.user.role_status.delegate')
+// i18n-tasks-use t('enums.user.role_status.senior_delegate')
+// i18n-tasks-use t('enums.user.role_status.member')
+// i18n-tasks-use t('enums.user.role_status.senior_member')
+// i18n-tasks-use t('enums.user.role_status.leader')
 
 export default function RolesTab({ userId }) {
+  const currentUser = useLoadedData(currentUserUrl);
   const {
     data, sync, loading, error,
   } = useLoadedData(roleListUrl(userId));
+  const { save, saving } = useSaveAction();
+
+  const [open, setOpen] = React.useState(false);
+  const [currentUserRoles, setCurrentUserRoles] = React.useState(null);
+  const [currentUserFetchError, setCurrentUserFetchError] = React.useState(null);
+
+  const activeNonHiddenRoles = React.useMemo(
+    () => {
+      if (data) {
+        return data.activeRoles.filter((role) => role.group.is_hidden === false);
+      }
+      return [];
+    },
+    [data],
+  );
+  const isDelegate = React.useMemo(
+    () => {
+      if (data) {
+        return data.activeRoles.some((role) => role.group.group_type === 'delegate_regions');
+      }
+      return false;
+    },
+    [data],
+  );
+
+  React.useEffect(() => {
+    if (!currentUser.loading) {
+      save(roleListUrl(currentUser.data.id), {}, (roles) => {
+        setCurrentUserRoles(roles.activeRoles.reduce((roleMap, role) => ({
+          ...roleMap,
+          [role.group.id]: role,
+        }), {}));
+      }, { method: 'GET', body: null }, setCurrentUserFetchError);
+    }
+  }, [currentUser.data, currentUser.loading, save]);
```

Comment:
I'm confused why we're `save`ing things in a side effect here. I would expect the only saving to be done in `RoleForm` when the (current) user makes a change.

And why we can't get the current user roles via `useLoadedData` -- is it (just) because the url isn't known until `currentUser` finishes loading, or is there something else too?

Reply:
Hi @kr-matthews, we are not actually saving here, instead we are fetching the current logged in user's roles. For rendering this page, we will be needing two data: roles of the user whose page is edited, and roles of the logged-in user. But we won't be having loggedIn user's ID, that has to be fetched initially. Once we have that, we will fetch the loggedIn user's roles.

Do you have any better suggestion to handle this? I'm thinking to pass `userId` as props to this component. I didn't do that because as per one of the previous discussions in another PR, we shouldn't pass any props to components rendered from rails, unless it's something like query params. But can `userId` can be considered as an exception as the future router can probably have `loggedInUserId` as it will be very much used in the whole application? What do you think? (@gregorbg would like to get your opinion as well on this)

Reply:
I'm confused when you say that we're not saving; the inside of the `if` starts with `save(roleListUrl(...`.

I don't really know enough to comment on `userId` being an exception. But if you don't want to pass `userId` as a prop then you can still make this work cleanly. Generally, if you have `a`, and need to fetch `b` and `c`, but you need `b` to get `c`, then you can make a wrapper component. So instead of
```
function MyComponent({ a }) {
  const b = useLoadedData(...)
  
  // an effect here to fetch c once b is loaded

  if (...anything is loading...) return <Loading />
  if (...anything is errored...) return <Error />

  // actual component return, given b and c have loaded
}
```
you can do
```
function MyComponent({ a }) {
  const b = useLoadedData()

  if (b.isLoading) return <Loading />
  if (b.isError) return <Error />

  return <HelperComponent a={a} b={b} />
}

function HelperComponent({ a, b}) {
  const c = useLoadedData(...something with b...)

  if (c.isLoading) return <Loading />
  if (c.isError) return <Error />

  // actual component return, given b and c have loaded
}
```

Reply:
I apologize if the previous discussion about this topic caused a mis-understanding. The rule of thumb is **not** that passing props from Rails in the `react_component` is forbidden. It is only forbidden/impractical/discouraged to pass data to React components that changes / re-loads during the lifecycle of the component.

For example, say you're writing a React component to display a list of upcoming competitions, and that component has pagination. The pagination makes it so that the list of competitions (or to be precise, the current page of upcoming competitions) is determined by what the user clicks in their browser at runtime. In that case, you shouldn't pass an initial batch of competitions to the component as props, nor should you pass a complete list of _all_ competitions to the React component and let React "butcher" the pagination by artificially chunking a pre-set array. You should instead load the data because it makes sense to load the data.

In this case, the user ID of the user whose roles are being edited does never change, so it's completely and utterly fine to pass it as prop.

Reply:
Okay, got it. Thanks @gregorbg. And sorry for making it confused @kr-matthews. I didn't think of using another component, that would have been a better approach if userId was not passed as props. Since userId is passed as props now, I guess this issue is solved.